images/builder/main.go

```diff
@@ -257,27 +269,69 @@ func mergeMaps(maps ...map[string]string) map[string]string {
 
 func parseFlags() options {
 	o := options{}
+	flag.BoolVar(&o.buildAtRoot, "build-at-root", false, "If true, uses the root of the workspace as the context for GCB run.")
 	flag.StringVar(&o.logDir, "log-dir", "", "If provided, build logs will be sent to files in this directory instead of to stdout/stderr.")
 	flag.StringVar(&o.scratchBucket, "scratch-bucket", "", "The complete GCS path for Cloud Build to store scratch files (sources, logs).")
 	flag.StringVar(&o.project, "project", "", "If specified, use a non-default GCP project.")
 	flag.BoolVar(&o.allowDirty, "allow-dirty", false, "If true, allow pushing dirty builds.")
 	flag.StringVar(&o.variant, "variant", "", "If specified, build only the given variant. An error if no variants are defined.")
-	flag.StringVar(&o.envPassthrough, "env-passthrough", "", "Comma-separated list of specified environment variables to be passed to GCB as subtitutions with an _ prefix. If the variable doesn't exist, the substitution will exist but be empty.")
+	flag.StringVar(&o.envPassthrough, "env-passthrough", "", "Comma-separated list of specified environment variables to be passed to GCB as substitutions with an _ prefix. If the variable doesn't exist, the substitution will exist but be empty.")
+
 	flag.Parse()
+
 	if flag.NArg() < 1 {
 		_, _ = fmt.Fprintln(os.Stderr, "expected an image directory to be provided")
 		os.Exit(1)
 	}
-	o.imageDirectory = flag.Arg(0)
+
+	o.imageDirectory = strings.TrimSuffix(flag.Arg(0), "/")
+
 	return o
 }
 
 func main() {
 	o := parseFlags()
-	if err := cdToRootDir(); err != nil {
-		log.Fatalf("Failed to cd to root: %v\n", err)
+
+	imageDirErr := o.validateImageDir()
+	if imageDirErr != nil {
+		log.Fatalf("Could not validate image directory: %v", imageDirErr)
 	}
 
+	if o.buildAtRoot {
+		o.cloudbuildPath = path.Join(o.imageDirectory, "cloudbuild.yaml")
+
+		if bazelWorkspace := os.Getenv("BUILD_WORKSPACE_DIRECTORY"); bazelWorkspace != "" {
+			if err := os.Chdir(bazelWorkspace); err != nil {
+				log.Fatalf("Failed to chdir to bazel workspace (%s): %v", bazelWorkspace, err)
+			}
+
+			cmd := exec.Command("git", "rev-parse", "--show-toplevel")
+			rootDir, err := cmd.Output()
+			if err != nil {
+				log.Fatalf("Failed to git rev-parse workspace: %v\n", err)
+
+				chdirErr := os.Chdir(strings.TrimSpace(string(rootDir)))
+				if chdirErr != nil {
+					log.Fatalf("Failed to cd to root of workspace: %v\n", chdirErr)
+				}
+			}
+		}
+	} else {
+		o.cloudbuildPath = "cloudbuild.yaml"
```

Comment:
I'm not sure that this `cloudbuildPath` option is what we want. There are essentially two concepts we have:

- Where are our config files located?
- What are we sending to GCB?

Notably, there are _two_ config files - `cloudbuild.yaml` and `variants.yaml` - which should be together, but not necessarily in the root. I think that this should be the path to that _directory_, and the paths to the relevant files can be assembled by appending to it later.

The other concept - what we send to GCB - either matches the config dir (by default), or is "the root", which we are really interpreting as "the current working directory" (which is a slightly awkward concept).

I still wonder whether it would make sense to instead of having `--build-at-root`, instead have `--upload-dir`, which would be equivalent to the current build-at-root with `--upload-dir=.` Then you would invoke the builder as `image-builder --upload-dir=. ./images/builder/` or something. I think in this case, both `upload-dir` and the position argument should be relative to the _current_ working directory, rather than eachother (since there is not actually any requirement that the config be part of the build).